File: C++/SetMatrixZeroes.cc
// Given a m x n matrix, if an element is 0, set its entire row and column to 0. Do it in place.

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        if (matrix.empty()) {
            return;
        }
        bool fst_row_zero = any_of(matrix[0].begin(), matrix[0].end(), [](int e){return e == 0;});
        for (int i = 1; i < matrix.size(); i++) {
            for (int j = 0; j < matrix[0].size(); j++) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = matrix[0][j] = 0;
                }
            }
        }

        for (int i = 1; i < matrix.size(); i++) {
            if (matrix[i][0] == 0) {
                fill(matrix[i].begin(), matrix[i].end(), 0);
            }
        }
        for (int i = 0; i < matrix[0].size(); i++) {
            if (matrix[0][i] == 0) {
                for_each(matrix.begin(), matrix.end(), [i](vector<int>& e){e[i] = 0;});
            }
        }
        if (fst_row_zero) {
            fill(matrix[0].begin(), matrix[0].end(), 0);
        }
    }
};

File: Java/SetMatrixZeroes.java
// Given a m x n matrix, if an element is 0, set its entire row and column to 0. Do it in place.

public class SetMatrixZeroes {
    public void setZeroes(int[][] matrix) {
        if (matrix.length == 0) {
            return;
        }
        int m = matrix.length;
        int n = matrix[0].length;

        boolean zeroCol0 = false;

        for (int i = 0; i < m; i++) {
            if (matrix[i][0] == 0) {
                zeroCol0 = true;
            }
            for (int j = 1; j < n; j++) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }

        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j > 0; j--) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
            if (zeroCol0) {
                matrix[i][0] = 0;
            }
        }
    }
}
